caffe2/operators/quantized/int8_average_pool_op.h

```diff
@@ -174,21 +47,100 @@ class Int8AveragePoolOp final : public ConvPoolOpBase<CPUContext> {
+            "failed to create QNNPACK Global Average Pooling operator");
+        CAFFE_ENFORCE(this->qnnpackGlobalOperator_ != nullptr);
+      }
+
+      const qnnp_status setupStatus = qnnp_setup_global_average_pooling_nwc_q8(
+        this->qnnpackGlobalOperator_,
+        X.t.dim32(0), X.t.dim32(1) * X.t.dim32(2),
+        X.t.template data<uint8_t>(), channels,
+        Y->t.template mutable_data<uint8_t>(), channels);
+      CAFFE_ENFORCE(
+          setupStatus == qnnp_status_success,
+          "failed to setup QNNPACK Global Average Pooling operator");
+
+#ifdef FBCODE_CAFFE2
+      const qnnp_status runStatus =
+          qnnp_run_operator(this->qnnpackGlobalOperator_,
+            nullptr /* thread pool */);
+#else
+      pthreadpool_t threadpool =
+          reinterpret_cast<pthreadpool_t>(ws_->GetThreadPool());
+      const qnnp_status runStatus =
+          qnnp_run_operator(this->qnnpackGlobalOperator_, threadpool);
+#endif
+      CAFFE_ENFORCE(
+          runStatus == qnnp_status_success,
+          "failed to run QNNPACK Global Average Pooling operator");
+    } else {
+      if (this->qnnpackOperator_ == nullptr) {
+        const qnnp_status createStatus = qnnp_create_average_pooling2d_nhwc_q8(
+          pad_t(), pad_r(), pad_b(), pad_l(),
+          kernel_h(), kernel_w(),
+          stride_h(), stride_w(),
+          channels,
+          X.zero_point, X.scale,
+          Y->zero_point, Y->scale,
+          activationLimits(Y->scale, Y->zero_point, Ac).first,
+          activationLimits(Y->scale, Y->zero_point, Ac).second,
+          &this->qnnpackOperator_);
+        CAFFE_ENFORCE(
+            createStatus == qnnp_status_success,
+            "failed to create QNNPACK Average Pooling operator");
+        CAFFE_ENFORCE(this->qnnpackOperator_ != nullptr);
+      }
+
+      const qnnp_status setupStatus = qnnp_setup_average_pooling2d_nhwc_q8(
+          this->qnnpackOperator_,
+          X.t.dim32(0), X.t.dim32(1), X.t.dim32(2),
+          X.t.template data<uint8_t>(), channels,
+          Y->t.template mutable_data<uint8_t>(), channels,
+          nullptr /* thread pool */);
+      CAFFE_ENFORCE(
+          setupStatus == qnnp_status_success,
+          "failed to setup QNNPACK Average Pooling operator");
+
+#ifdef FBCODE_CAFFE2
+      const qnnp_status runStatus =
+          qnnp_run_operator(this->qnnpackOperator_, nullptr /* thread pool */);
+#else
+      pthreadpool_t threadpool =
+          reinterpret_cast<pthreadpool_t>(ws_->GetThreadPool());
+      const qnnp_status runStatus =
+          qnnp_run_operator(this->qnnpackOperator_, threadpool);
+#endif
+      CAFFE_ENFORCE(
+          runStatus == qnnp_status_success,
+          "failed to run QNNPACK Average Pooling operator");
+    }
+
     return true;
   }
+ private:
+  // QNNPACK Average Pooling operator
+  qnnp_operator_t qnnpackOperator_{nullptr};
+  // QNNPACK Global Average Pooling operator
+  qnnp_operator_t qnnpackGlobalOperator_{nullptr};
```

Comment:
this could be the same as qnnpackOperator_ ?

Reply:
we may need both, because we consider non-global pooling with `kernel size == image size` as global max pooling. if the same operator is fed with images of different sizes, if may execute as global pooling for some images, and as non-global pooling for others